File: find_vowels/find_vowels.go
// You can edit this code!
// Click here and start typing.
package main

import (
	"fmt"
	"regexp"
)

func main() {
	test := "This is test string"
	fmt.Printf("Count by normal fn %d", countVowels(test))
	fmt.Printf("\n Count by Regx fn %d", countByRegx(test))

}

func countByRegx(str string) int {
	comp := regexp.MustCompile("(?i)[aeiou]")

	strLen := comp.FindAllString(str, -1)
	return len(strLen)

}

// Normal way by looping and checking count using this
func countVowels(str string) int {
	checker := []string{"a", "i", "e", "o", "u"}
	count := 0
	for _, c := range str {
		var lower string
		if c >= 'A' && c <= 'Z' {
			lower = string(c + 32)
		} else {
			lower = string(c)
		}
		if Contains(checker, string(lower)) {
			count++
		}
	}
	return count
}

// Contains checks if a slice contains an element, using generics
func Contains[T comparable](slice []T, elem T) bool {
	for _, t := range slice {
		if t == elem {
			return true
		}
	}
	return false
}

File: find_vowels/find_vowels.rs
use regex::Regex;

fn main() {
    let test = "This is test string";
    println!("Count by normal fn {}", count_vowels(test));
    println!("Count by regex fn {}", count_by_regex(test));
}

fn count_by_regex(str: &str) -> usize {
    let regex = Regex::new(r"(?i)[aeiou]").unwrap();
    let matches = regex.find_iter(str).count();
    matches
}

fn count_vowels(str: &str) -> usize {
    let vowels = ['a', 'e', 'i', 'o', 'u'];
    let mut count = 0;

    for c in str.chars() {
        let lower = c.to_ascii_lowercase();
        if vowels.contains(&lower) {
            count += 1;
        }
    }

    count
}

// Rust's `Vec` has a built-in `contains` method
fn contains<T: PartialEq>(slice: &[T], elem: &T) -> bool {
    slice.contains(elem)
}
